src/services/SearchService.ts

```diff
@@ -1,7 +1,281 @@
+      if (await this.checkAvailableDates(i.id, dates)) {
+        needed.add(i.id);
+      }
+    }
+
+    return Array.from(needed);
+  }
+
+  private getAskDates(): DateTime[] {
+    if (!this.ask.checkIn || !this.ask.checkOut) {
+      throw ApiError.BadRequest('invalid dates in ask');
+    }
+
+    const checkIn = DateTime.fromFormat(this.checkInTime, 'hhmm', {
+      zone: this.facility.policies?.timezone
+    }).setZone('utc');
+
+    let from = DateTime.fromObject(
+      {
+        ...this.ask.checkIn,
+        hour: checkIn.hour,
+        minute: checkIn.minute
+      },
+      { zone: 'utc' }
+    );
+    const to = DateTime.fromObject(
+      {
+        ...this.ask.checkOut,
+        hour: checkIn.hour,
+        minute: checkIn.minute
+      },
+      { zone: 'utc' }
+    );
+
+    const dates: DateTime[] = [];
+
+    while (from <= to) {
+      dates.push(from);
+      from = from.plus({ days: 1 });
+    }
+
+    return dates;
+  }
+
+  private async checkRules(
+    dates: DateTime[],
+    spaceId: string
+  ): Promise<boolean> {
+    const spaceRuleRepository = new SpaceRuleRepository(
+      this.facilityId,
+      spaceId
+    );
+
+    const noticeRequirementRule = (await this.getRule(
+      'notice_required',
+      spaceRuleRepository
+    )) as NoticeRequiredRule;
+
+    const firstDay = dates[0];
+
+    const checkIn = DateTime.now().setZone('utc');
+    const interval = Interval.fromDateTimes(checkIn, firstDay).toDuration([
+      'days'
+    ]);
+
+    const intervalSeconds = convertDaysToSeconds(interval.get('days'));
+
+    if (
+      noticeRequirementRule &&
+      noticeRequirementRule.value > intervalSeconds
+    ) {
+      return false;
+    }
+
+    const formattedFirstDay = firstDay.toFormat('ccc').toLowerCase();
+    const lOSRule = (await this.getRule(
+      'length_of_stay',
+      spaceRuleRepository,
+      formattedFirstDay
+    )) as DayOfWeekLOSRule;
+
+    const lOS = dates.length;
+    let minLOS = 0;
+    let maxLOS = 0;
+
+    try {
+      minLOS = lOSRule[formattedFirstDay]?.minLengthOfStay || 0;
+    } catch (e) {
+      //rule not exist
+    }
+
+    try {
+      maxLOS = lOSRule[formattedFirstDay]?.maxLengthOfStay || 0;
+    } catch (e) {
+      //rule not exist
+    }
+
+    return (minLOS === 0 || lOS > minLOS) && (maxLOS === 0 || lOS < maxLOS);
+  }
+
+  private async getRule(
+    ruleName: RulesItemKey,
+    spaceRuleRepository: SpaceRuleRepository,
+    weekDay: null | string = null
+  ) {
+    let rule = await spaceRuleRepository.getRule(ruleName);
+
+    if (!rule || (weekDay && !(weekDay in rule))) {
+      rule = await this.facilityRuleRepository.getRule(ruleName);
+    }
+
+    return rule;
+  }
+
+  private checkSuitableQuantity(space: Space) {
+    const numOfAdults =
+      space.maxNumberOfAdultOccupantsOneof.oneofKind ===
+      'maxNumberOfAdultOccupants'
+        ? space.maxNumberOfAdultOccupantsOneof.maxNumberOfAdultOccupants
+        : 0;
+
+    const numOfChildren =
```

Comment:
Renaming to `maxNumOfChildren` to avoid confusing with `ask`

Reply:
fixed